### DISCUSSION TITLE
Register Functionality?

### DISCUSSION THREAD

**Discussion by OfflinePing** [pocketbase/pocketbase#4201] [Q&A]
Hello,
I want to ask a couple of questions

My Code:
```golang
type User struct {
	ID       string `db:"id" json:"id"`
	Username string `db:"username" json:"username"`
	Email    string `db:"email" json:"email"`
	Password string `db:"password" json:"password"`
	Name     string `db:"name" json:"name"`
	Verified bool   `db:"verified" json:"verified"`
	Avatar   string `db:"avatar" json:"avatar"`
}

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodPost,
			Path:   "/register",
			Handler: func(c echo.Context) error {
				var user User
				if err := c.Bind(&user); err != nil {
					return err
				}

				collection, err := app.Dao().FindCollectionByNameOrId("users")
				if err != nil {
					return err
				}

				record := models.NewRecord(collection)

				record.Set("username", user.Username)
				record.Set("email", user.Email)
				record.Set("password", user.Password)
				record.Set("passwordConfirm", user.Password)

				if err := app.Dao().SaveRecord(record); err != nil {
					return err
				}

				return c.NoContent(http.StatusOK)
			},
			Middlewares: []echo.MiddlewareFunc{
				apis.ActivityLogger(app),
			},
		})
		return nil
	})

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodGet,
			Path:   "/GetUsers",
			Handler: func(c echo.Context) error {
				var Users []User
				err := app.Dao().DB().NewQuery("SELECT * FROM users").All(&Users)
				if err != nil {
					return err
				}
				return c.JSON(http.StatusOK, Users)
			},
			Middlewares: []echo.MiddlewareFunc{
				apis.RequireAdminAuth(),
				apis.ActivityLogger(app),
			},
		})
		return nil
	})

	if err := app.Start(); err != nil {
		panic(err)
	}
}
```

1. So right now I have done it via my own Endpoint but do I even have to make my own?
2. This does work but why can I only post once to the API and then it reports this:
```
ERROR POST /register
└─ Something went wrong while processing your request.
   └─ UNIQUE constraint failed: users.tokenKey
```
4. Is there a way to send the Verification E-Mail via a Function or do I need to call the API for that?

**Comment by ganigeorgiev** [comment]
> So right now I have done it via my own Endpoint but do I even have to make my own?

If you are asking for whether you need your own registration function - most of the time no as the default collection Create action should be sufficient (see the "API Preview" code sample in the Admin UI example for your collection or check the generic [web API docs](https://pocketbase.io/docs/api-records/#create-record)).


> This does work but why can I only post once to the API and then it reports this

The function is not the correct way to create a user record because there are no "password" or "passwordConfirm" fields in the model db record. The above fields are available only as request params and will work only if you use the [`forms.RecordUpsert`](https://pocketbase.io/docs/go-records/#create-new-record-with-data-validations). If you want to create an user record by directly operating with `models.Record` (aka. directly interacting with the underlying db representation) you can check the example in the migrations https://pocketbase.io/docs/go-migrations/#creating-new-auth-record/

> Is there a way to send the Verification E-Mail via a Function or do I need to call the API for that?

Yes, you can call [`mails.SendRecordVerification(app, record)`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.20.7/mails#SendRecordVerification).
